### PR TITLE
[Bugfix] [Relay] fix a bug of  printing dataflow pattern

### PR DESCRIPTION
**Pull request by kfeng123**
When recursion of dataflow patterns is used, the pattern graph may not be a DAG. If recursion is encountered, ReprPrint of dataflow pattern may fall in a dead loop.
This PR solves the bug of dataflow pattern printing, and is the first PR for the [pre-RFC](https://discuss.tvm.apache.org/t/on-the-applications-of-the-composition-of-dataflow-patterns/15344?u=kfeng123).

### CODE REVIEW

**Comment by kfeng123** [comment]
cc @masahi @mbrookhart

**Comment by masahi** [comment]
btw, does the dataflow pattern support recursion at all?

File: include/tvm/relay/dataflow_pattern.h [hunk 1]
```diff
@@ -537,6 +539,41 @@ DFPattern IsTuple(const Array<DFPattern>& fields);
...
 DFPattern IsTupleGetItem(const DFPattern tuple, int index = -1);
 
+/*! \brief A printer class to print pattern. */
+class DFPatternPrinter : public ReprPrinter {
+ public:
+  std::stringstream string_stream{};
+
+  std::unordered_map<DFPattern, std::pair<size_t, std::string>, ObjectPtrHash, ObjectPtrEqual>
+      memo_{};
+  std::vector<DFPattern> recursed_patterns{};
```
Comment (masahi) [inline include/tvm/relay/dataflow_pattern.h:549]: Please document what you mean by "recursed_patterns". Probably "recursive_patterns" would be more correct.

File: include/tvm/relay/dataflow_pattern.h [hunk 2]
```diff
@@ -537,6 +539,41 @@ DFPattern IsTuple(const Array<DFPattern>& fields);
...
+
+inline std::ostream& operator<<(std::ostream& os,
+                                const DFPattern& n) {  // NOLINT(*)
+  std::stringstream string_stream{}, tmp_stream{};
+  DFPatternPrinter printer{tmp_stream};
+  printer.Print(n);
+  string_stream << "Main pattern is:" << std::endl;
+  string_stream << printer.string_stream.str();
+  string_stream << std::endl;
+  string_stream << "Auxiliary patterns are:";
```
Comment (masahi) [inline include/tvm/relay/dataflow_pattern.h:566]: Remove "is" and "are" above

**Comment by kfeng123** [comment]
> btw, does the dataflow pattern support recursion (loop) at all?

Surprisingly, the answer is yes. Indeed, I was thinking to add the support for recursion, only to find that the existing dataflow pattern language has alreadly (silently) support it.
But currently recursion may only implemented in C++, not python.
The following is an example taken from the [pre-RFC](https://discuss.tvm.apache.org/t/on-the-applications-of-the-composition-of-dataflow-patterns/15344?u=kfeng123) doc:

``` python
TVM_REGISTER_GLOBAL("relay.dataflow_pattern.my_pattern")
    .set_body_typed([]() {
    DFPattern dense_pattern = IsOp("nn.dense")({IsWildcard(), IsWildcard()});
    ObjectPtr<CallPatternNode> the_pattern_ptr = make_object<CallPatternNode>();
    the_pattern_ptr->op = IsOp("cast");
    the_pattern_ptr->args.clear();
    CallPattern the_pattern = CallPattern(the_pattern_ptr);
    AltPattern or_pattern{the_pattern, dense_pattern};
    the_pattern_ptr->args.push_back(or_pattern);
    //LOG(INFO) << PrettyPrint(the_pattern); // TODO: BUG!
  return the_pattern;
});
```
This simple pattern matches a nn.dense followed by an arbitrary number of cast. You can test this pattern via the following python code:

``` python
class TheRewrite(DFPatternCallback):
    def __init__(self):
        super(TheRewrite, self).__init__(rewrite_once = True)
        pattern = tvm.get_global_func("relay.dataflow_pattern.my_pattern")()
        self.pattern = pattern
    def callback(self, pre, post, node_map):
        return relay.nn.relu(post)

mod = create_model() # define a model
the_rewrite = TheRewrite()
out = rewrite(the_rewrite, mod["main"])
```

Another application of recursion is PR [#15362](https://github.com/apache/tvm/pull/15362), which I do not know how to achieve without recursion. That PR is useful, and can really improve the computational graph for some quantized models.

I would like to examine the pattern matching code further in the following days.
If I find any bug with pattern recursion, I would like to raise PRs.

File: include/tvm/relay/dataflow_pattern.h (hunk 2 above)
Comment (kfeng123) [reply include/tvm/relay/dataflow_pattern.h]: Done

File: include/tvm/relay/dataflow_pattern.h (hunk 1 above)
Comment (kfeng123) [reply include/tvm/relay/dataflow_pattern.h]: Documentation added. To be consistent with the printed text, I choose to name this variable to be auxiliary_patterns

**Comment by kfeng123** [comment]
@tvm-bot rerun

**Review by masahi — APPROVED** [review]

**Pull request merged**